cores/arduino/stm32/rtc.c

```diff
@@ -0,0 +1,492 @@
+      __HAL_RCC_LSEDRIVE_CONFIG(RCC_LSEDRIVE_LOW);
+#endif
+      RCC_OscInitStruct.OscillatorType =  RCC_OSCILLATORTYPE_LSE;
+      RCC_OscInitStruct.PLL.PLLState = RCC_PLL_NONE;
+      RCC_OscInitStruct.LSEState = RCC_LSE_ON;
+      if(HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
+        Error_Handler();
+      }
+    }
+
+    PeriphClkInit.PeriphClockSelection = RCC_PERIPHCLK_RTC;
+    PeriphClkInit.RTCClockSelection = RCC_RTCCLKSOURCE_LSE;
+    if(HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK) {
+      Error_Handler();
+    }
+  } else if(source == HSE_CLOCK) {
+    // Enable the clock if not already set by user.
+    if(__HAL_RCC_GET_FLAG(RCC_FLAG_HSERDY) == RESET) {
+      RCC_OscInitStruct.OscillatorType =  RCC_OSCILLATORTYPE_HSE;
+      RCC_OscInitStruct.PLL.PLLState = RCC_PLL_NONE;
+      RCC_OscInitStruct.HSEState = RCC_HSE_ON;
+      if(HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
+        Error_Handler();
+      }
+    }
+
+    PeriphClkInit.PeriphClockSelection = RCC_PERIPHCLK_RTC;
+#if defined(STM32F1xx)
+    PeriphClkInit.RTCClockSelection = RCC_RTCCLKSOURCE_HSE_DIV128;
+#elif defined(STM32L4xx) || defined(STM32F0xx) || defined(STM32F3xx)
+    PeriphClkInit.RTCClockSelection = RCC_RTCCLKSOURCE_HSE_DIV32;
+#else
+    PeriphClkInit.RTCClockSelection = RCC_RTCCLKSOURCE_HSE_DIV8;
+#endif
+    if(HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK) {
+      Error_Handler();
+    }
+  } else {
+    // Enable the clock if not already set by user.
+    if(__HAL_RCC_GET_FLAG(RCC_FLAG_LSIRDY) == RESET) {
+      RCC_OscInitStruct.OscillatorType =  RCC_OSCILLATORTYPE_LSI;
+      RCC_OscInitStruct.PLL.PLLState = RCC_PLL_NONE;
+      RCC_OscInitStruct.LSIState = RCC_LSI_ON;
+      if(HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
+        Error_Handler();
+      }
+    }
+
+    PeriphClkInit.PeriphClockSelection = RCC_PERIPHCLK_RTC;
+    PeriphClkInit.RTCClockSelection = RCC_RTCCLKSOURCE_LSI;
+    if(HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK) {
+      Error_Handler();
+    }
+  }
+
+  __HAL_RCC_RTC_ENABLE();
+}
+
+/**
+  * @brief RTC Initialization
+  *        This function configures the RTC time and calendar. By default, the
+  *        RTC is set to the 1st January 2017 0:0:0:00
+  * @param format: enable the RTC in 12 or 24 hours mode
+  * @retval None
+  */
+void RTC_init(hourFormat_t format, sourceClock_t source)
+{
+  initFormat = format;
+
+  // Set RTC clock
+  RTC_setClock(source);
+
+  RtcHandle.Instance = RTC;
+
+#if defined(STM32F1xx)
+  RtcHandle.Init.AsynchPrediv = RTC_AUTO_1_SECOND; //Leave HAL calculate the prescaler
+  RtcHandle.Init.OutPut = RTC_OUTPUTSOURCE_NONE;
+  UNUSED(format);
+#else
+  if(format == HOUR_FORMAT_12) {
+    RtcHandle.Init.HourFormat = RTC_HOURFORMAT_12;
+  } else {
+    RtcHandle.Init.HourFormat = RTC_HOURFORMAT_24;
+  }
+  RtcHandle.Init.OutPut = RTC_OUTPUT_DISABLE;
+  if(source == LSE_CLOCK) {
+    RtcHandle.Init.AsynchPrediv = LSE_ASYNCH_PREDIV;
```

Comment:
[HSE|LSE|LSI]_A?SYNC_PREDIV should not be hard-coded. User should have a way to dynamically change them without add a dedicated define. At least an API to change the default value.
Anyway, as mentioned in rtc.h, it should be possible to define the best PREDIV